rfd/0095-application-access-okta-integration.md

```diff
@@ -0,0 +1,315 @@
+basics are more ironed out.
+
+A user will be able to submit access requests to specific applications and groups through the
+API or UI. These requests will submit access requests through Teleport's
+[existing access request functionality](https://goteleport.com/docs/access-controls/access-requests/).
+The Okta service will monitor these approval requests and take appropriate
+action based on the request and the resource targeted.
+
+### APIs used by the Okta service
+
+The Okta provider will use the following APIs:
+
+* [List applications](https://developer.okta.com/docs/reference/api/apps/#list-applications) for
+  listing all applications known to Okta. These results are paginated and may require multiple
+  calls.
+* [List groups](https://developer.okta.com/docs/reference/api/groups/#list-groups) for listing
+all groups known to Okta. These results are paginated and may require multiple calls.
+
+### Backend modifications
+
+#### Request access service
+
+The Okta service will need to actively monitor and take action on access requests, which is a
+mechanism which does not currently exist today.
+
+### Audit events
+
+A number of new audit events will be created as part of this effort:
+
+| Event Name | Description |
+|------------|-------------|
+| `OKTA_GROUPS_UPDATE` | Emitted when groups synchronized from Okta have changed. |
+| `OKTA_APPLICATIONS_UPDATE` | Emitted when applications synchronized from Okta have have changed. |
+| `OKTA_SYNC_FAILURE` | Emitted when an Okta synchronization attempt fails. |
+| `OKTA_ACCESS_REQUEST_APPROVED` | Emitted when a user request for an Okta resource was approved. |
+| `OKTA_ACCESS_REQUEST_DENIED` | Emitted when a user request for an Okta resource was denied. |
+
+### Security
+
+* It's possible to capture sensitive data in our internal Teleport database. This could be
+  undesirable for customers or other users of Teleport. We should take care that the sensitive
+  identity provider data (like Okta API tokens) are never communicated to or stored outside of
+  the configuration file.
+* The mapping of Okta users to the appropriate Teleport user is a concern. If a user has been
+  mismapped to an Okta user, it will be possible for that user to request access on behalf of
+  that user. Administrators should take care with their Okta identity mappings, or even better,
+  ensure that Teleport logins map directly to Okta users.
+
+### Implementation plan
+
+#### Okta service configuration
+
+Implement the ability to configure the Okta service. Doing this first will make subsequent
+testing and verification easier as we'll be able to pass in API URLs and API tokens.
+Part of this will include implementing any stubs needed for the Okta service itself.
+
+#### Okta service API communication
+
+The Okta service will be able to communicate with Okta and retrieve lists of
+applications and groups without doing anything with them yet.
+
+####
```

Comment:
Unfinished?

Reply:
Oh huh. I filled all this out, but I guess it didn't save? Sorry about that.